src/librustc/middle/check_const.rs

```diff
@@ -813,8 +815,25 @@ impl<'a, 'tcx> euv::Delegate<'tcx> for CheckCrateVisitor<'a, 'tcx> {
                 }
                 mc::cat_deref(ref cmt, _, _) |
                 mc::cat_downcast(ref cmt, _) |
-                mc::cat_interior(ref cmt, _) => cur = cmt,
+                mc::cat_interior(ref cmt, _) => {
+                    is_interior = true;
+                    cur = cmt;
+                }
 
+                mc::cat_rvalue(ty::ReStatic) if is_interior => {
+                    match self.mode {
+                        Mode::Static | Mode::StaticMut => {
+                            span_err!(
+                                self.tcx.sess,
+                                consume_span,
+                                E0398,
+                                "cannot dereference a &'static T",
```

Comment:
in case you have

```rust
static A: i32 = 5;
static B: i32 = *&A;
```

which is forbidden, but

```rust
static A: i32 = 5;
let x = *&A;
```

should be allowed

but as I said. this also breaks dereferencing references to constants